### PR TITLE
feat: Migrate away from Struts front-end [DHIS2-14092]

### PR DESCRIPTION
**Pull request by netroms**
## Summary:

**WARNING: This is PR is intended for development purposes only, until all the web apps are migrated to use the new non Struts equivalent endpoints.**

This PR removes the need to use Struts as the front-end layer completely.
The PR makes it possible to start a fully functional DHIS2 server with back-end and front-end, with the use of an embedded Jetty server serving the static content.

**Note: This PR does not remove any features, so the old/normal way of starting the server still works.**

When the front-end apps are fully migrated and the login app is done, we will make this into a production ready feature.

### How to start the server from IntelliJ:
1. Compile the project fully with Maven (We need to do this to download all the web apps and package the static front-end files resources, this only needs to be done once.)
2. Open the JettyEmbeddedCoreWeb.java and right-click the play icon on the main() method.
3. Modify the run configuration and set the DHIS2_HOME env. variable. (No need for this if you already have a config in the default location/folder)
4. Run the Application configuration. (Debug mode is preferred, so hot reload will work)

### How to start the server from the command line:
1. Compile the project fully with Maven (We need to do this to download all the web apps and package the static front-end files resources, this only needs to be done once.)
2. Run the _run-api.sh_ script.

### Caveats/workarounds: 
This PR makes use of a couple of standard servlets, running outside the Spring context, in order to make it possible to override certain endpoints as a workaround until the apps are migrated away from using the Struts endpoints.
These servlets are defined in the JettyEmbeddedCoreWeb class.
The servlets are: GetModulesServlet and RootPageServlet.

The GetModulesServlet emulates the _/dhis-web-commons/menu/getModules.action_ endpoint used for the menu. This servlet will make an internal request to the _/api/webModules_, which is the new endpoint replacing the old Struts GetModulesAction.
This servlet is temporary, and will be removed when front end apps are migrated to the new API endpoint.

The RootPageServlet servlet's use is to serve the login page, which now is only a static .html file, called login.html
This is just a copy of the source of the rendered login page and serves as a temporary solution until the new login app is finished. This static login page only supports password login. No other features like: reset password, self registration, two-factor and OIDC login is currently supported. 

### Apps not working:
* Old data entry app

### Bonus:
When running the server inside IntelliJ in debug mode, it supports hot reload of classes, just right click in the editor and click: “Compile and reload file”

### Jira:
[DHIS2-14092](https://dhis2.atlassian.net/browse/DHIS2-14092)

### CODE REVIEW

**Review by larshelge — APPROVED** [review]

File: dhis-2/dhis-web-embedded-jetty/src/main/java/org/hisp/dhis/web/embeddedjetty/JettyEmbeddedCoreWeb.java [hunk 1]
```diff
@@ -42,14 +42,15 @@
...
+import org.springframework.web.context.request.RequestContextListener;
 import org.springframework.web.context.support.AnnotationConfigWebApplicationContext;
 import org.springframework.web.filter.CharacterEncodingFilter;
 import org.springframework.web.filter.DelegatingFilterProxy;
 import org.springframework.web.servlet.DispatcherServlet;
 
 public class JettyEmbeddedCoreWeb extends EmbeddedJettyBase
 {
-    private static final int DEFAULT_HTTP_PORT = 9080;
+    private static final int DEFAULT_HTTP_PORT = 8080;
```
Comment (larshelge) [inline dhis-2/dhis-web-embedded-jetty/src/main/java/org/hisp/dhis/web/embeddedjetty/JettyEmbeddedCoreWeb.java:53]: Are we changing the default Jetty embedded port here?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/appmanager/WebModuleManager.java [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+
+/**
+ * @author Morten Svanæs <msvanaes@dhis2.org>
+ */
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class WebModuleManager
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/appmanager/WebModuleManager.java:56]: Should we change all references from `WebModule` to `WebApp`?

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/WebModulesController.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+import org.springframework.web.bind.annotation.*;
+
+/**
+ * @author Morten Svanæs <msvanaes@dhis2.org>
+ */
+
+@OpenApi.Tags( "webmodules" )
+@RestController
+@AllArgsConstructor
+@RequestMapping( "/webModules" )
```
Comment (larshelge) [inline dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/WebModulesController.java:61]: Change to `/webApps`?

File: dhis-2/dhis-web-embedded-jetty/src/main/java/org/hisp/dhis/web/embeddedjetty/GetModulesServlet.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+            Authentication authentication = context.getAuthentication();
+            CurrentUserDetailsImpl principal = (CurrentUserDetailsImpl) authentication.getPrincipal();
+            String username = principal.getUsername();
+
+            resp.setContentType( "application/json" );
+            resp.setStatus( HttpServletResponse.SC_OK );
+
+            RequestDispatcher dispatcher = getServletContext().getRequestDispatcher(
+                "/api/webModules" + "?username=" + username );
```
Comment (larshelge) [inline dhis-2/dhis-web-embedded-jetty/src/main/java/org/hisp/dhis/web/embeddedjetty/GetModulesServlet.java:71]: Should we get rid of the "module" word and instead use "app"?

**Review by jbee — APPROVED** [review]

File: dhis-2/dhis-web-embedded-jetty/src/main/java/org/hisp/dhis/web/embeddedjetty/EmbeddedJettyBase.java [hunk 5]
```diff
@@ -37,11 +37,12 @@
...
 import lombok.extern.slf4j.Slf4j;
 
-import org.eclipse.jetty.server.Connector;
-import org.eclipse.jetty.server.HttpConfiguration;
-import org.eclipse.jetty.server.HttpConnectionFactory;
-import org.eclipse.jetty.server.Server;
-import org.eclipse.jetty.server.ServerConnector;
+import org.eclipse.jetty.rewrite.handler.RedirectPatternRule;
+import org.eclipse.jetty.rewrite.handler.RewriteHandler;
+import org.eclipse.jetty.server.*;
```
Comment (jason-p-pickering) [inline dhis-2/dhis-web-embedded-jetty/src/main/java/org/hisp/dhis/web/embeddedjetty/EmbeddedJettyBase.java:42]: Should this be fixed/optimized?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/appmanager/AppMenuManager.java [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.appmanager;
+
+import static org.hisp.dhis.util.FileUtils.getResourceFileAsString;
+
+import java.io.IOException;
+import java.util.*;
```
Comment (jason-p-pickering) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/appmanager/AppMenuManager.java:33]: Optimize imports.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/system/DefaultSystemService.java [hunk 7]
```diff
@@ -32,11 +32,7 @@
 import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
-import java.util.Calendar;
-import java.util.Date;
-import java.util.List;
-import java.util.Properties;
-import java.util.TimeZone;
+import java.util.*;
```
Comment (jason-p-pickering) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/system/DefaultSystemService.java:35]: Optimize imports.

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/AppController.java [hunk 8]
```diff
@@ -63,14 +65,7 @@
...
 import org.springframework.stereotype.Controller;
-import org.springframework.web.bind.annotation.DeleteMapping;
-import org.springframework.web.bind.annotation.GetMapping;
-import org.springframework.web.bind.annotation.PathVariable;
-import org.springframework.web.bind.annotation.PostMapping;
-import org.springframework.web.bind.annotation.PutMapping;
-import org.springframework.web.bind.annotation.RequestMapping;
-import org.springframework.web.bind.annotation.RequestParam;
-import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.*;
```
Comment (jason-p-pickering) [inline dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/AppController.java:68]: Should you use the wildcard here?

File: dhis-2/dhis-web/dhis-web-commons/src/main/java/org/hisp/dhis/webportal/module/DefaultModuleManager.java [hunk 9]
```diff
@@ -27,16 +27,7 @@
...
-import java.util.Collection;
-import java.util.Collections;
-import java.util.Comparator;
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Locale;
-import java.util.Map;
-import java.util.Set;
+import java.util.*;
```
Comment (jason-p-pickering) [inline dhis-2/dhis-web/dhis-web-commons/src/main/java/org/hisp/dhis/webportal/module/DefaultModuleManager.java:30]: Again, wildcard imports.

**Pull request merged**